<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείγματος θα εξετάσετε μία υλοποίηση με Q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>

Ο αλγόριθμος Q-Learning ανήκει στην κατηγορία του Reinforcement Learning και πιο συγκεκριμένα εντάσσεται στην υποκατηγορία του **model-free learning**, όπου ο agent **δεν** προσπαθεί να μάθει ρητά τις environment state transition και reward functions, αλλά καταλήγει σε ένα βέλτιστο policy/πολιτική από τις αλληλεπιδράσεις του με το περιβάλλον. Ουσιαστικά, ο τρόπος που το κάνει αυτό είναι μαθαίνοντας καλά και κακά actions μέσω trial and error.

Ο Q-Learning αλγόριθμος προσεγγίζει τη συνάρτηση Q-function από τα δείγματα του $Q(s,a)$ που παρατηρούνται κατά την αλληλεπίδραση του agent με το environment.

Ο αλγόριθμος έχει ως εξής σε ψευδοκώδικα:
- Ο πίνακας $Q(s,a)$ αρχικοποιείται τυχαία
- Για κάθε interaction του agent με το environment:
    - παρατηρείται το reward του action του $r(s,a)$ και το state transition (νέο state $s'$)
    - υπολογίζεται το observed Q-value $Q_{o b s}(s, a)=r(s, a)+\gamma \max _{a^{\prime}} Q\left(s^{\prime}, a^{\prime}\right)$
    - ανανέωση της τιμής του Q-function $Q(s,a)$ ως εξής $Q(s, a)=(1-\eta) Q(s, a)+\eta \cdot Q_{o b s}(s, a)$

όπου $0< \eta \leq 1$ είναι το learning rate (ρυθμός μάθησης) και $0 \leq \gamma \leq 1$ είναι το discount factor όπως είχε περιγραφεί και στο εργαστήριο με τους Value και Policy Iteration αλγορίθμους. Αυτός ο παράγοντας δίνει μικρότερα βάρη σε μελλοντικά rewards. Θέτοντας τιμή 0 στο $\gamma$ ουσιαστικά κάνουμε τον agent να ενδιαφέρεται μόνο για άμεσα (short-term) rewards, δηλαδή να είναι greedy, ενώ αν θέσουμε τιμή ίση με 1 τότε ο agent δίνει ίδιο βάρος σε short-term και long-term rewards.

Το Reinforcement Learning βρίσκεται ενδιάμεσα στο Supervised και Unsupervised Learning. Ο στόχος στο Reinforcement Learning είναι η εύρεση της βέλτιστης ακολουθίας αποφάσεων που επιτρέπουν σε έναν agent να λύσει ένα πρόβλημα μεγιστοποίησης μακροπρόθεσμου reward. Αυτή η ακολουθία actions μαθαίνεται μέσω αλληλεπίδρασης του agent με το environment και την παρατήρηση των rewards σε κάθε state.
Για παράδειγμα, παιχνίδια όπως το Go, είναι κλασικά παραδείγματα εφαρμογής Reinforcement Learning.

Η βασική διαφορά του Q-learning από το Value and Policy iteration αλγορίθμους έγκειται στο γεγονός ότι ο Value and Policy Iteration.

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [1]:
!pip install cmake 'gym[atari]' scipy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [3]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [4]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [30]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

Δημιουργώντας το environment, δημιουργείται και ένα dictionary $P$ με keys τα states (κωδικοποίηση 0 έως 499 για τα 500 πιθανά states συνολικά) και values ένα dictionary για κάθε state, το οποίο με τη σειρά του έχει ως keys τα δυνατά actions (κωδικοποίηση 0 έως 5 για τα 6 πιθανά actions συνολικά) και ως values μια λίστα με μία tuple μέσα. Αρχικά οι τιμές που δημιουργούνται είναι κάποιες αυθαίρετες default values.

Άρα για ένα επιλεγμένο state ως key στο dictionary παίρνουμε ως value το εξής:

{action0: [(probability0, nextstate0, reward0, done0)],<br>
action1: [(probability1, nextstate1, reward1, done1)],<br>
action2: [(probability2, nextstate2, reward2, done2)],<br>
action3: [(probability3, nextstate3, reward3, done3)],<br>
action4: [(probability4, nextstate4, reward4, done4)],<br>
action5: [(probability5, nextstate5, reward5, done5)]}

Τα στοιχεία της κάθε tuple με τη σειρά είναι:
- Η πιθανότητα να συμβεί το intended action του agent
- Το next state id (0-499)
- Tο reward του agent για το action αυτό
- Ένα boolean flag αν είναι done ή όχι μια διαδρομή, δηλαδή αν επιτυχώς ο επιβάτης του ταξί έφτασε στον προορισμό του.

Το mapping των actions έχει ως εξής:
- 0: south
- 1: north
- 2: east
- 3: west
- 4: pickup
- 5: dropoff

ενώ η πιθανότητα είναι πάντα 1, δηλαδή ο τρόπος με τον οποίο συμβαίνουν οι κινήσεις του agent είναι deterministic. Αυτό σημαίνει ότι αν ένας agent αποφασίσει να κάνει ένα action, πχ να κινηθεί προς τα πάνω, τότε με $100\%$ πιθανότητα θα κινηθεί προς τα πάνω. Ακόμη, όλες οι κινήσεις έχουν ως reward $-1$, ώστε να προσπαθήσει ο agent να ελαχιστοποιήσει τον αριθμό κινήσεών του προς την πλήρωση του σκοπού του, ενώ οι pickup/dropoff actions έχουν reward $-10$ αν δεν είναι οι σωστές τοποθεσίες για pickup/dropoff του πελάτη του ταξί, ενώ αν είναι σωστό το action του dropoff (σωστό dropoff location/state), τότε το reward είναι $+20$. Κάθε επιτυχημένο dropoff ορίζει το τέλος ενός episode.

Τέλος, να σημειώσουμε πως η κίνηση προς κάποιον τοίχο συνεπάγεται παραμονή στο ίδιο κελί/state του ταξί και reward -1, πράγμα που οδηγεί τον agent στο να κινείται και να μην παραμένει σε ένα κελί.



<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

In [31]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 3038
Penalties incurred: 987


Το παραπάνω cell κώδικα, επιλέγει τυχαία όταν βρίσκεται σε κάθε state, ένα action από τα 6 πιθανά actions του agent, τα οποία είναι τα εξής:
- 0 = south
- 1 = north
- 2 = east
- 3 = west
- 4 = pickup
- 5 = dropoff

Με βάση το τυχαία επιλεγμένο action οδηγείται σε ένα νέο state ενώ λαμβάνει και ένα reward. 

Παρατηρούμε ότι ο agent χρειάζεται χιλιάδες βηματα για να καταφέρει να παραδώσει κάποιον πελάτη του ταξί στην σωστή τοποθεσία, ενώ στο μεταξύ κάνει σχεδόν 1000 penalties (!), δηλαδή κάνει ισάριθμες φορές λάθος στο pickup/dropoff του πελάτη.

Αυτό συμβαίνει γιατί ο agent δε μαθαίνει και λειτουργεί με τυχαίο τρόπο. Αντιθέτως, με τη χρήση του Q-Learning, ο agent αποκτά μνήμη και μπορεί να μάθει με τρόπο trial and error από τις προηγούμενες προσπάθειές του μαθαίνοντας τελικά την καλύτερη action για κάθε state ώστε να μεγιστοποιεί το reward του.

<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

Η παράμετρος **α** είναι το learning rate (παραπάνω δόθηκε ως **η** στις εξισώσεις του Q-function), δηλαδή ο ρυθμός με τον οποίο ανανεώνεται το Q-value σε κάθε βήμα. Τιμή 1 σημαίνει ότι ανανεώνεται η τιμή του Q για κάθε state και action χωρίς να υπολογίζεται η προηγούμενη τιμή του Q-table (πίνακας με γραμμές τα states και columns τα actions). Γενικώς, το learning rate είναι μια παράμετρος που δείχνει πόση γνώση συγκρατεί από την προηγούμενη εμπειρία του ο agent σε κάθε βήμα.

Η παράμετρο **γ** μοντελοποιεί την ανάθεση βαρών σε μελλοντικά rewards. Θέτοντας τιμή 0 στο **γ** ουσιαστικά κάνουμε τον agent να ενδιαφέρεται μόνο για άμεσα (short-term) rewards, δηλαδή να είναι greedy, ενώ αν θέσουμε τιμή ίση με 1 τότε ο agent δίνει ίδιο βάρος σε short-term και long-term rewards.

In [7]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [8]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 1min 5s, sys: 10.8 s, total: 1min 16s
Wall time: 1min 8s


In [9]:
q_table[328]

array([ -2.40991108,  -2.27325184,  -2.40616307,  -2.36013082,
       -10.60593154, -10.6527113 ])

In [10]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties = 0, 0
episodes = 100

for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)

        if reward == -10:
            penalties += 1

        epochs += 1

    total_penalties += penalties
    total_epochs += epochs

print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")

Results after 100 episodes:
Average timesteps per episode: 12.91
Average penalties per episode: 0.0


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

Σε 100 επεισόδια:
Average timesteps per episode: 12.91
Average penalties per episode: 0.0

| Measure |Random agent's performace values| Q-learning agent's performance   values|   
|---|---|---|
| Μέσος αριθμός παραβάσεων ανά επεισόδιο  | 987 | 0.0 |  
| Μέσος αριθμός βημάτων ανά διαδρομή  | 3038 | 12.91  |   
| Μέσος αριθμός ανταμοιβών ανά κίνηση  | $\frac{987 \cdot (-10) + (3038 - 987 - 1) \cdot (-1) + 20}{3038} = -3.917$  |  $\frac{0 \cdot (-10) + (12.91 - 0 - 1) \cdot (-1) + 20}{12.91} = \frac{-11.91+20}{12.91} = 0.627$  |   

Είναι εμφανές, ότι ο Q learning agent είναι σαφώς καλύτερος από τον random καθώς χρειάζεται λιγότερα από 13 βήματα κατά μέσο όρο για να μεταφέρει καθέναν από τους 100 πελάτες του ταξί και δεν κάνει λανθασμένα actions που καταλήγουν σε reward $-10$.

# Information sources:
- <a href="https://www.learndatasci.com/tutorials/reinforcement-q-learning-scratch-python-openai-gym/">Q-Learning - OpenAI gym taxi</a>
- <a href="https://medium.com/@m.alzantot/deep-reinforcement-learning-demysitifed-episode-2-policy-iteration-value-iteration-and-q-978f9e89ddaa">Value and policy iteration for the Frozen Lake problem</a>